arroyo/processing/strategies/dead_letter_queue/policies/abstract.py

```diff
@@ -1,5 +1,53 @@
+import base64
 from abc import ABC, abstractmethod
-from typing import Any, Sequence
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Mapping, Optional, Sequence, Union
+
+Serializable = Union[str, bytes]
+
+
+@dataclass(frozen=True)
+class InvalidMessage:
+    """
+    A dataclass to generally represent a bad Kafka message.
+    The Kafka specific fields (offset etc) are all optional
+    since this can be used to pass almost anything to the
+    DLQ.
+
+    A `reason` can be provided for debugging purposes. If a
+    produce policy is configured on the relevant DLQ, a message
+    in the form returned by `to_dict()` will be produced via
+    the policy.
+
+    An `InvalidMessages` exception should be raised containing
+    one or more objects of this class in order to actually
+    pass data to the DLQ.
+    """
+
+    payload: Serializable
+    timestamp: datetime
+    reason: Optional[str] = None
+    consumer_group: Optional[str] = None
+    partition: Optional[int] = None
+    offset: Optional[int] = None
+
+    def to_dict(self) -> Mapping[str, str]:
+        if isinstance(self.payload, bytes):
+            try:
+                decoded = self.payload.decode("utf-8")
+            except UnicodeDecodeError:
+                decoded = base64.b64encode(self.payload).decode("utf-8")
```

Comment:
I think you should be consistent in how you encode the data here. Otherwise sometimes it will be b64 encoded and sometimes not. Maybe add a flag or prefix so we know that we are seeing the b64 version of the data.

Reply:
Added a prefix.